### PR TITLE
#166272104 Password Reset

### PR DESCRIPTION
**Pull request by mosinmiloluwa01**
#### What does this PR do?
This PR implements password reset feature

#### Description of Task to be completed?
- write tests for recover email and reset password features
- implement recover email feature
- implement reset password feature
- create middleware for validation

#### How should this be manually tested?
- input "http://localhost:3000/api/v1/auth/recoverEmail" post route on postman
- supply the user's email
- click on send
- go to your email, click on the "recover password" button 
- the button takes you to the browser, copy the link and input in postman
- input a new password and confirm password
- click on send

#### Any background context you want to provide?
User must have node, postgres and postman installed

#### What are the relevant pivotal tracker stories?
[166272104](https://www.pivotaltracker.com/story/show/166272104)

### CODE REVIEW

File: server/tests/controller/recoverEmail.spec.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+const { expect } = chai;
+
+describe('Recover Email', () => {
+  it('should not send email with an invalid email format', (done) => {
+    chai
+      .request(app)
+      .post('/api/v1/auth/recoverEmail')
+      .send({ email: '12' })
+      .end((err, res) => {
+        expect(res.status).to.be.equal(404);
```
Comment (Eazybee) [inline server/tests/controller/recoverEmail.spec.js:13]: you might want to test more than just status code

File: server/tests/controller/recoverEmail.spec.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        done(err);
+      });
+  });
+  it('should not send email with an incorrect email', (done) => {
```
Comment (Eazybee) [inline server/tests/controller/recoverEmail.spec.js:19]: new line needed here for readability

File: server/tests/controller/recoverEmail.spec.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      .request(app)
+      .post('/api/v1/auth/recoverEmail')
+      .send({ email: 'a@b.com' })
+      .end((err, res) => {
+        expect(res.status).to.be.equal(404);
+        expect(res.body.message).to.equal('This email does not exist');
+        done(err);
+      });
+  });
+  it('should not send email with a correct email', (done) => {
```
Comment (Eazybee) [inline server/tests/controller/recoverEmail.spec.js:31]: same here

File: server/tests/controller/resetPassword.spec.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    chai
+      .request(app)
+      .patch(`/api/v1/auth/resetPassword?token=${token}`)
+      .send({ password: 'abcfegt569', confirmPassword: 'abcfegt56' })
+      .end((err, res) => {
+        expect(res.status).to.be.equal(403);
+        done(err);
+      });
+  });
+  it('should not reset password if password has an incorrect format', (done) => {
```
Comment (Eazybee) [inline server/tests/controller/resetPassword.spec.js:25]: same here

File: server/tests/controller/resetPassword.spec.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    chai
+      .request(app)
+      .patch(`/api/v1/auth/resetPassword?token=${token}`)
+      .send({ password: '569', confirmPassword: '569' })
+      .end((err, res) => {
+        expect(res.status).to.be.equal(400);
+        done(err);
+      });
+  });
+  it('should not send email with a correct email', (done) => {
```
Comment (Eazybee) [inline server/tests/controller/resetPassword.spec.js]: same here

**Review by Eazybee — DISMISSED** [review]

**Review by akinyeleolat — APPROVED** [review]

**Review by EmmanuelSage — APPROVED** [review]

File: server/emailTemplates/recoverEmail.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        text: 'Reset Password',
+        link
+      }
+    },
+    outro: 'If you did not initiate this request, please ignore this mail.'
+  }
+});
+
+const generateEmail = (email, req) => {
+  const token = tokenGenerator('noid', false, process.env.TOKEN_EXPIRY_DATE, process.env.SECRET, email);
```
Comment (benfluleck) [inline server/emailTemplates/recoverEmail.js]: What is noid ?

Comment (benfluleck) [reply server/emailTemplates/recoverEmail.js]: why is `isVerified` set to false for a rest email ?

File: server/emailTemplates/recoverEmail.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        color: '#B02091',
+        text: 'Reset Password',
+        link
+      }
+    },
+    outro: 'If you did not initiate this request, please ignore this mail.'
+  }
+});
+
+const generateEmail = (email, req) => {
```
Comment (benfluleck) [inline server/emailTemplates/recoverEmail.js]: shouldn't this be named `generateResetEmail`

File: server/helpers/utils.js [hunk 8]
```diff
@@ -36,10 +36,14 @@ export const toLowerCaseAndTrim = (inputObject) => {
  * @param {string} isVerified
  * @param {string} tokenExpiryDate
  * @param {string} secret
+ * @param {string} email
  * @return {string} token
  */
-export const tokenGenerator = (id, isVerified, tokenExpiryDate = '1h', secret = 'secret') => {
+export const tokenGenerator = (id, isVerified, tokenExpiryDate = '1h', secret = 'secret', email) => {
   const payload = { id, isVerified };
+  if (email && toLowerCaseAndTrim(email) !== '') {
```
Comment (benfluleck) [inline server/helpers/utils.js]: What is this line doing? Why not simply `if(email)`

**Review by benfluleck — CHANGES_REQUESTED** [review]
Changes requested also rebase

**Review by benfluleck — APPROVED** [review]

**Pull request merged**